wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationParticipantsService.swift

```diff
@@ -0,0 +1,243 @@
+
+        do {
+            try await internalAddParticipants(users, to: conversation)
+        } catch let error as FederationError {
+            try await handleFederationError(
+                error,
+                users: users,
+                conversation: conversation
+            )
+        }
+    }
+
+    private func internalAddParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation
+    ) async throws {
+
+        let messageProtocol = await context.perform { conversation.messageProtocol }
+
+        switch messageProtocol {
+        case .proteus:
+
+            try await proteusParticipantsService.addParticipants(users, to: conversation)
+
+        case .mls:
+
+            guard let mlsParticipantsService else {
+                throw ConversationParticipantsError.missingMLSParticipantsService
+            }
+
+            do {
+                try await mlsParticipantsService.addParticipants(users, to: conversation)
+            } catch MLSConversationParticipantsError.ignoredUsers(users: _) {
+                // TODO: Insert system message
+                // To be done in https://wearezeta.atlassian.net/browse/WPB-2228
+            }
+
+        case .mixed:
+
+            guard let mlsParticipantsService else {
+                throw ConversationParticipantsError.missingMLSParticipantsService
+            }
+
+            try await proteusParticipantsService.addParticipants(users, to: conversation)
+
+            // For mixed protocol we only try once and don't handle errors
+            try? await mlsParticipantsService.addParticipants(users, to: conversation)
+        }
+    }
+
+    private func validate(
+        users: [ZMUser],
+        conversation: ZMConversation
+    ) async throws {
+        try await context.perform { [self] in
+            guard
+                conversation.conversationType == .group,
+                !users.isEmpty,
+                !users.contains(ZMUser.selfUser(in: context))
+            else {
+                throw ConversationParticipantsError.invalidOperation
+            }
+        }
+    }
+
+    private func handleFederationError(
+        _ error: FederationError,
+        users: [ZMUser],
+        conversation: ZMConversation
+    ) async throws {
+        switch error {
+        case .unreachableDomains(let domains):
+            let unreachableUsers = await context.perform { users.belongingTo(domains: domains) }
+
+            if unreachableUsers.isEmpty {
+
+                /// If there are no users from unreachable domains, this means that the backend tried and failed to check for non-fully connected graphs
+                /// because some of the existing participants are currently unreachable.
+                /// As a result, we should inform that users can't be added and not retry the request.
+                /// https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/822149401/Non-fully+connected+federation+graphs
```

Comment:
I think we could simplify this comment with something like:

```
// Backend is not able to determine which users are unreachable.

Reply:
done https://github.com/wireapp/wire-ios/pull/696/commits/b95a5350ed59908e8d9578ebf12399ddb197650f